routing/control_tower_test.go

```diff
@@ -0,0 +1,253 @@
+)
+
+var (
+	priv, _ = btcec.NewPrivateKey(btcec.S256())
+	pub     = priv.PubKey()
+
+	testHop = &route.Hop{
+		PubKeyBytes:      route.NewVertex(pub),
+		ChannelID:        12345,
+		OutgoingTimeLock: 111,
+		AmtToForward:     555,
+	}
+
+	testRoute = route.Route{
+		TotalTimeLock: 123,
+		TotalAmount:   1234567,
+		SourcePubKey:  route.NewVertex(pub),
+		Hops: []*route.Hop{
+			testHop,
+			testHop,
+		},
+	}
+)
+
+// TestPaymentControlSubscribeUnknown tests that subscribing to an unknown
+// payment fails.
+func TestPaymentControlSubscribeUnknown(t *testing.T) {
+	t.Parallel()
+
+	db, err := initDB()
+	if err != nil {
+		t.Fatalf("unable to init db: %v", err)
+	}
+
+	pControl := NewControlTower(channeldb.NewPaymentControl(db))
+
+	// Subscription should fail when the payment is not known.
+	_, err = pControl.SubscribePayment(lntypes.Hash{1})
+	if err != channeldb.ErrPaymentNotInitiated {
+		t.Fatal("expected subscribe to fail for unknown payment")
+	}
+}
+
+// TestPaymentControlSubscribeSuccess tests that payment updates for a
+// successful payment are properly sent to subscribers.
+func TestPaymentControlSubscribeSuccess(t *testing.T) {
+	t.Parallel()
+
+	db, err := initDB()
+	if err != nil {
+		t.Fatalf("unable to init db: %v", err)
+	}
+
+	pControl := NewControlTower(channeldb.NewPaymentControl(db))
+
+	// Initiate a payment.
+	info, attempt, preimg, err := genInfo()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	err = pControl.InitPayment(info.PaymentHash, info)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Subscription should succeed and immediately report the InFlight
+	// status.
+	subscriber1, err := pControl.SubscribePayment(info.PaymentHash)
+	if err != nil {
+		t.Fatalf("expected subscribe to succeed, but got: %v", err)
+	}
+
+	// Register an attempt.
+	err = pControl.RegisterAttempt(info.PaymentHash, attempt)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Register a second subscriber after the first attempt has started.
+	subscriber2, err := pControl.SubscribePayment(info.PaymentHash)
+	if err != nil {
+		t.Fatalf("expected subscribe to succeed, but got: %v", err)
+	}
+
+	// Mark the payment as successful.
+	if err := pControl.Success(info.PaymentHash, preimg); err != nil {
+		t.Fatal(err)
+	}
+
+	// Register a third subscriber after the payment succeeded.
+	subscriber3, err := pControl.SubscribePayment(info.PaymentHash)
+	if err != nil {
+		t.Fatalf("expected subscribe to succeed, but got: %v", err)
+	}
+
+	// We expect all subscribers to now report the final outcome followed by
+	// no other events.
+	for _, s := range []chan PaymentResult{
+		subscriber1, subscriber2, subscriber3,
+	} {
+		event := <-s
+		if !event.Success {
+			t.Fatal("unexpected payment state")
+		}
+		if event.Preimage != preimg {
+			t.Fatal("unexpected preimage")
+		}
+		if !reflect.DeepEqual(event.Route, &attempt.Route) {
+			t.Fatal("unexpected route")
+		}
+
+		// After the final event, we expect the channel to be closed.
+		select {
+		case _, ok := <-s:
```

Comment:
does the go runtime guarantee that the close event will arrive immediately?

Reply:
Good point, added timeout